LSM: Implement Tree.lookup()

ifreund opened the pull request:
~~This is implemented on top of #85, so that PR should be reviewed/merged first.~~

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -2593,7 +2725,7 @@ pub fn main() !void {
     _ = tree.put;
     _ = tree.remove;
-    _ = tree.lookup;
+    _ = tree.lookup_value;
```
Let's keep this as `lookup()` and then rather use `context` or something like that instead of `lookup` as the local variable within the `Lookup` struct.

ifreund replied:
Fine by me, my thought was to mirror `Grid.read_block()` and `Grid.Read` but I also agree that `lookup()` is the nicer name to expose outside of the tree.

ifreund replied:
renamed `lookup_value()` -> `lookup()` and `lookup: *Lookup` to `context: *LookupContext`.

ifreund committed:
LSM: Implement Tree.lookup()

jorangreef commented on src/lsm/grid.zig:
```diff
@@ -71,6 +68,9 @@ pub fn GridType(comptime Storage: type) type {
     return struct {
         const Grid = @This();
 
+        pub const BlockPtr = *align(config.sector_size) [block_size]u8;
+        pub const BlockPtrConst = *align(config.sector_size) const [block_size]u8;
```
Neat to have this in Grid.

ifreund replied:
Glad you agree, I thought these made more sense there rather than in `Table` :)

jorangreef commented on src/lsm/table.zig:
```diff
@@ -684,27 +683,55 @@ pub fn TableType(
             );
         }
 
+        pub inline fn index_data_addresses_const(index_block: BlockPtrConst) []const u64 {
```
Good to have a `const` version.

Let's rename this to `index_data_addresses_used()` which returns const, and only returns the slice that is used, excluding the padding. This way we get more checked arithmetic safety at the call site e.g. in `read_index_callback()`.

jorangreef replied:
This can also use `index_data_addresses()` internally, to have a single code path. Same also for the other related `_const()` methods.

ifreund replied:
> This can also use index_data_addresses() internally, to have a single code path. Same also for the other related _const() methods.

Actually it can't, because `index_data_addresses()` takes a mutable pointer. Zig doesn't have a good solution to this problem yet, I think we'll eventually want to add a `@constCast()` builtin but not sure where andrew stands on that. SpexGuy things it's a good idea iirc. One can cast away const using `@ptrToInt()`/`@intToPtr()` currently but that makes code harder for LLVM to understand/optimize.

ifreund replied:
> Let's rename this to index_data_addresses_used() which returns const, and only returns the slice that is used, excluding the padding. This way we get more checked arithmetic safety at the call site e.g. in read_index_callback().

I'm going ahead with reworking all of the `_const` versions I added in this way

jorangreef commented on src/lsm/table.zig:
```diff
@@ -759,13 +786,17 @@ pub fn TableType(
+                Value,
+                data_block[data.values_offset..][0..data.values_size],
+            );
+            return values[0..used];
```
Is there any advantage to duplicating the `data_block_values()` implementation here, since the function is already inline?

ifreund replied:
`data_block_values()` takes a mutable pointer so it can't be used here.

Perhaps I should have called this function `data_block_values_used_const()` but we don't have any need for mutable version currently...

jorangreef replied:
Ah thanks, I missed the pointer to const in the function signature, looking only at the implementation.

Yes, `data_block_values_used()` being implicitly const is fine, it doesn't need to be part of the name, since it's meant for reading what was written, and what was written is always immutable because everything is copy-on-write.

jorangreef replied:
We could also use `const slice` here like the others.

ifreund replied:
> We could also use `const slice` here like the others.

done!

jorangreef commented on src/lsm/table.zig:
```diff
@@ -775,9 +806,51 @@ pub fn TableType(
+                        key,
+                    );
+                    if (result.exact) {
+                        return &values[result.index];
```
Would you please add a verification check on the value key here, to ensure that it is the value we were looking for, if `config.verify` is set.

And then do the same for the `null` path, after merging them to a single `return null`, i.e. to do a plain linear search on all the values, without using the Eytzinger layout or binary search, to double-check that we didn't miss the value if it exists. Our Eyztinger and binary search are both well tested, but this would test also the interaction between them.

jorangreef replied:
`config.verify` is like Tom Holland's “Training Wheels Protocol”. 😜

jorangreef commented on src/lsm/table.zig:
```diff
@@ -775,9 +806,51 @@ pub fn TableType(
+            return filter_block[filter.filter_offset..][0..filter.filter_size];
+        }
+
+        pub fn data_block_search(data_block: BlockPtrConst, key: Key) ?*const Value {
```
Would you do a small fuzz test for this? It's such an important function, and it's also nice and contained, depending only on the table layout.

jorangreef replied:
For the fuzz test, round-tripping here could work well and also give coverage of Builder. As an example:

1. For value type, create 1000 random values, plus boolean array.
2. Select N of these 1000 values, sort and build a table, marking values as inserted in the boolean array.
3. Test that all inserted values are found, according to the block they should be in, and that non-inserted values are not.
4. Go back to 2. and select a different subset, sometimes exploring very few small tables or tables at max capacity.
5. Go back to 1. and select a different sized Value, to have coverage for our different alignments.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
+
+            context.* = .{
+                .tree = tree,
+                .completion = undefined,
```
Would it make sense to have `completion` default to `undefined`, since it's only used later when kicking off a read?

ifreund replied:
I don't have any strong preference either way, this is just the style we've been using everywhere else for this pattern (journal, grid, etc.).

I think I started doing it this way because I found it a bit more obvious what was going on with completion at the usage site.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
+
+                assert(info.visible(context.it.snapshot));
+                assert(compare_keys(context.it.key, info.key_min) != .lt);
+                assert(compare_keys(context.it.key, info.key_max) != .gt);
```
It would be good to make the `info` the subject here, compared against the constant context key, since we are asserting that the iterator returned the correct info relative to the key.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
+
+            fn read_next_table(context: *LookupContext) void {
+                assert(context.current_table == null);
+                const info = context.it.next() orelse {
```
```suggestion
                const table = context.it.next() orelse {
```

Since that's what we do elsewhere whenever we work with a `TableInfo`.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
+                callback(context, value);
+            }
+
+            fn read_next_table(context: *LookupContext) void {
```
```suggestion
            fn read_index_block(context: *LookupContext) void {
```

ifreund replied:
I actually prefer the current name here as this function doesn't read the index block of the current table, it advances the iterator to the next table and reads the index block of that.

Even if we switch to storing the results of the manifest iterator in an array on Lookup initialization, this function will be responsible for advancing the index into that array.

ifreund replied:
Ended up splitting off an `advance_to_next_level()` function and doing this rename in e8c99fae7acbd7222064ff4cedc4bd9234f0a1cf. I had forgotten that the first call wouldn't need to advance the index and this ended up being cleaner.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
+                );
+            }
+
+            fn read_index_callback(completion: *Grid.Read, index: Grid.BlockPtrConst) void {
```
```suggestion
            fn read_index_block_callback(completion: *Grid.Read, index_block: Grid.BlockPtrConst) void {
```

Also for reading filter and data blocks.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
+            fn read_index_callback(completion: *Grid.Read, index: Grid.BlockPtrConst) void {
+                const context = @fieldParentPtr(LookupContext, "completion", completion);
+                const data_block_index = Table.index_data_block_for_key(index, context.it.key);
+                const filter_block_index = data_block_index / Table.filter.data_block_count_max;
```
```suggestion
                const filter_block_index = @divFloor(data_block_index, Table.filter.data_block_count_max);
```

To show the reader we've thought through the ramifications of any rounding error.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
+                };
+
+                context.tree.grid.read_block(
+                    read_filter_callback,
```
It's going to be awesome when this is running hot through the cache.

ifreund replied:
Indeed, I really like how straightforward this control flow is :)

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
+
+            fn read_index_callback(completion: *Grid.Read, index: Grid.BlockPtrConst) void {
+                const context = @fieldParentPtr(LookupContext, "completion", completion);
+                const data_block_index = Table.index_data_block_for_key(index, context.it.key);
```
The `_index` suffix here overloads onto the `index` of `index_block`.

Would be great to have a helper to encapsulate this, i.e. to return the filter block checksum and address, or data block checksum and address in one shot, so the caller doesn't need to also index into addresses/checksums etc.

ifreund replied:
`index_data_block_for_key()` does a binary search so we only want to call it once. A helper that returns all 4 values from the index might be a good idea though.

ifreund replied:
Implemented in a8557545921569eb3794258ef2cb2bfb46fe4ad9

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
+
+            fn read_filter_callback(completion: *Grid.Read, filter_block: Grid.BlockPtrConst) void {
+                const context = @fieldParentPtr(LookupContext, "completion", completion);
+
```
Let's assert everything that we expect of the context here. While we do assert that the `current_table` is not null below, that's only in one of the branches, not both.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
+            fn read_filter_callback(completion: *Grid.Read, filter_block: Grid.BlockPtrConst) void {
+                const context = @fieldParentPtr(LookupContext, "completion", completion);
+
+                const filter_bytes = Table.filter_block_filter_const(filter_block);
```
Do we need to restrict the filter slice here to only that which is used, or do we always use the full piece of memory?

ifreund replied:
The code is correct as is. We don't use the full block as filter memory but we always pass the full memory used for the filter to the functions in `bloom_filter.zig`.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
+                        context.current_table.?.data_block_checksum,
+                    );
+                } else {
+                    // The key is not present in this table, move on.
```
```suggestion
                    // The key is not present in this table, check the next level.
```

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
+                data_block_checksum: u128,
+            } = null,
+
+            callback: fn (*Tree.LookupContext, ?*const Value) void,
```
What do you think about making this an optional, so that `null` can be used as a guard against any double call?

ifreund replied:
I don't think that would help in this case. We set the entire struct to `undefined` in `finish()` which will also protect against a double call and is more idiomatic.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
+
+            fn read_data_callback(completion: *Grid.Read, data_block: Grid.BlockPtrConst) void {
+                const context = @fieldParentPtr(LookupContext, "completion", completion);
+
```
Would be good to assert everything we expect of the `context` here too.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
+                if (Table.data_block_search(data_block, context.it.key)) |value| {
+                    context.finish(value);
+                } else {
+                    // The key is not present in this table, move on.
```
```suggestion
                    // The key is not present in this table, check the next level.
```

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -228,32 +235,119 @@ pub fn TreeType(comptime Table: type) type {
+            current_table: ?struct {
+                data_block_address: u64,
+                data_block_checksum: u128,
+            } = null,
```
Since we can no longer use static allocation when doing compactions, because we don't know how many tables in level B may be involved, short of the whole level, we are forced to update the manifest while the compaction is running. This would invalidate any existing iterators.

It's not a bug here yet, because we keep a clear separation between phases: `prefetch / commit / compact` and don't pipeline them (yet).

However, I wouldn't mind having the manifest iteration done in one shot here, to produce a single array of `config.lsm_levels` tables with `address/checksum` that we iterate using an index here. It's a little extra work for keys in the top levels, although 90% of keys are in the last level, and this means we just do constant work, which is not a bad thing.

For compaction, and other uses of the manifest, I want to try and keep all iterators internal to the manifest, so there's no risk of an iterator being invalidated, and so that when it comes to pipelining the commit and compaction phases, to overlap them (to hide the compaction latency behind the commit latency), we have everything in place already.

jorangreef replied:
We could do this with something like `manifest.table_for_key(snapshot, key, level) ?*const TableInfo`, then copying only the checksum and address for each level.

ifreund replied:
Implemented this in e8c99fae7acbd7222064ff4cedc4bd9234f0a1cf. I used the existing manifest iterator in a while loop instead of introducing new manifest API but it would be easy to swap that out if we change the public Manifest API.

jorangreef replied:
Great.

jorangreef reviewed (commented):
This was great reading @ifreund — nice work!

Pretty cool to be seeing the read path come together like this.

ifreund committed:
LSM: Add config.verify checks to Table.data_block_search()

ifreund committed:
LSM: Improve Tree.lookup() code style based on review

ifreund committed:
LSM: Iterate Manifest up front in Tree.lookup()

This makes it impossible for the Manifest iterator to be invalidated
during a lookup. This is a bit more expensive for keys that are near
the top of the LSM tree, but since ~90% of the keys are in the last
level this constant overhead should be acceptable.

This doesn't solve any current iterator invalidation bug since we don't
allow compaction and lookup to happen concurrently. Pipelining these is
something we may explore in the future however.

ifreund committed:
LSM: Add exhaustive assertions to lookup() path callbacks

jorangreef commented on src/lsm/table.zig:
```diff
@@ -684,27 +680,59 @@ pub fn TableType(
+                u64,
+                index_block[index.data_addresses_offset..][0..index.data_addresses_size],
+            );
+            return data_addresses[0..index_data_blocks_used(index_block)];
```
```suggestion
            const slice = mem.bytesAsSlice(
                u64,
                index_block[index.data_addresses_offset..][0..index.data_addresses_size],
            );
            return slice[0..index_data_blocks_used(index_block)];
```

To thin out the function body, and be consistent across all the `used()` methods.

For example, `index_data_keys_used()` currently has `keys` instead of `data_keys` whereas the others would have the `data_` prefix.

However, these `used()` functions also already repeat the subject three times, e.g. `keys` via `data_keys_offset`, `data_keys_size` and in the function name itself, so using something generic like `slice` is less visual overhead, puts the spotlight on the offset/size variables at play, and works across these functions consistently.

jorangreef commented on src/lsm/table.zig:
```diff
@@ -745,27 +773,52 @@ pub fn TableType(
+
+        /// Returns all data stored in the index block relating to a given key.
+        /// May be called on an index block only when the key is already in range of the table.
+        pub inline fn index_key_info(index_block: BlockPtrConst, key: Key) IndexKeyInfo {
```
```suggestion
        pub inline fn index_blocks_for_key(index_block: BlockPtrConst, key: Key) IndexBlocks {
```

Or `index_filter_and_data_block_for_key() IndexFilterAndDataBlock`.

jorangreef commented on src/lsm/table.zig:
```diff
@@ -745,27 +773,52 @@ pub fn TableType(
+                .filter_block_address = index_filter_addresses_used(index_block)[filter_block_i],
+                .filter_block_checksum = index_filter_checksums_used(index_block)[filter_block_i],
+                .data_block_address = index_data_addresses_used(index_block)[data_block_i],
+                .data_block_checksum = index_data_checksums_used(index_block)[data_block_i],
```
```suggestion
            const d = Table.index_data_block_for_key(index_block, key);
            const f = @divFloor(d, filter.data_block_count_max);

            return .{
                .filter_block_address = index_filter_addresses_used(index_block)[f],
                .filter_block_checksum = index_filter_checksums_used(index_block)[f],
                .data_block_address = index_data_addresses_used(index_block)[d],
                .data_block_checksum = index_data_checksums_used(index_block)[d],
```

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -227,32 +234,166 @@ pub fn TreeType(comptime Table: type) type {
+                    index_block_checksums[index_block_count] = table.checksum;
+                }
             }
+
```
What do we expect if `index_block_count==0` at the end of the loop?

I think we could return early here without the cost of computing the fingerprint if we know that no table in any level contains the key.

ifreund replied:
Good catch, this is actually a bug introduced in the refactor to do iteration up front. Right now we'll end up using an undefined address/checksum to try and read an index block from disk. With this change and the `index_block_count > 0` assert we'll have the bug fixed and be protected against re-introducing it.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -227,32 +234,166 @@ pub fn TreeType(comptime Table: type) type {
+            key: Key,
+            fingerprint: bloom_filter.Fingerprint,
+
+            index_block_current: u8 = 0,
```
We could go with just `index_block` here since that's the focus of the `LookupContext`, iterating the index blocks.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -227,32 +234,166 @@ pub fn TreeType(comptime Table: type) type {
+            fn read_index_block(context: *LookupContext) void {
+                assert(context.data_block == null);
+                assert(context.index_block_current < context.index_block_count);
+                assert(context.index_block_count <= config.lsm_levels);
```
We could add `assert(context.index_block_count > 0)` here since we don't expect to be reading blocks if that is ever 0.

jorangreef commented on src/lsm/tree.zig:
```diff
@@ -227,32 +234,166 @@ pub fn TreeType(comptime Table: type) type {
+                context.data_block = null;
+                context.read_index_block();
+            }
+        };
```
Great name. Perfect control flow.

ifreund replied:
Thanks, I was pretty happy with how this turned out :)

jorangreef reviewed (commented):
Great work @ifreund — this is ready to land once the test is in.

ifreund committed:
LSM: Improve Tree.lookup() code style based on review

ifreund committed:
LSM: Fix uninitalized read in Tree.lookup()

Since a recent refactor, the case where the key is not present anywhere
in the manifest is not properly handled. Add an early return to handle
this case and assertions to avoid re-introducing this bug.

ifreund committed:
LSM: Implement Table.Builder.init() and deinit()

ifreund committed:
LSM: Don't depend on Grid in Table.Builder

Instead pass in the newly acquired address and cluster as options in the
{data,filter,index}_block_finish() functions, using a struct argument to
avoid mixing up the order at the call site.

This allows isolated testing of the Builder type without instantiating a
Grid, Superblock, Storage and IO.

ifreund committed:
LSM: Make Table type not dependant on Storage

This is no longer used in the Table implementation, which makes sense.
Table deals purely with the data layout, not with writing that data
layout to disk.

ifreund commented:
Merging this now to avoid growing conflicts after the latest refactoring commits. The fuzz test for `Table.Builder` and `Table.data_block_search()` will come in a follow up commit and likely include `TableImmutable` as well.

Thanks for the reviews @jorangreef!
